Improve julia queries

mroavi opened the pull request.

mroavi committed:
Make tabs consistent, remove trailing whitespace

mroavi committed:
Highlight optional arguments

mroavi commented:
Before:
![image](https://user-images.githubusercontent.com/19517248/103176528-c005a880-4872-11eb-8c37-51030cb4f310.png)
After:
![image](https://user-images.githubusercontent.com/19517248/103176542-d1e74b80-4872-11eb-96a2-92535ff2cdc2.png)

mroavi committed:
Capture the "end" keyword as a number inside a range expr

mroavi commented:
Before:
![image](https://user-images.githubusercontent.com/19517248/103177453-f7785300-487a-11eb-9797-e4f6bfe7f4fd.png)
After:
![image](https://user-images.githubusercontent.com/19517248/103177463-052dd880-487b-11eb-88fa-355eee72116b.png)

mroavi commented:
The previous commit doesn't take care of all cases. If, for example, the `end` keyword is part of another expression, e.g. `end-1`, then it doesn't capture it. How can I express using this tree-sitter language that `end` should be captured at any level within a `range_expression`?
Edit: I think I figured it out. Using a *wildcard* node.

mroavi committed:
Fix: use a wild card node to match "end" in deeper exprs

mroavi commented:
Before:
![image](https://user-images.githubusercontent.com/19517248/103177929-d3b70c00-487e-11eb-8229-dd5b6479e8df.png)
After:
![image](https://user-images.githubusercontent.com/19517248/103177940-e2052800-487e-11eb-9c80-19d49318a213.png)

mroavi committed:
Add query for while statement

mroavi commented:
Before:
![image](https://user-images.githubusercontent.com/19517248/103179755-7aa3a400-488f-11eb-99a1-9c86f46849b1.png)
After:
![image](https://user-images.githubusercontent.com/19517248/103179764-8f803780-488f-11eb-82f6-2dd8870406ac.png)

mroavi committed:
Catch parameters that have parametric types

mroavi commented:
Before:
![image](https://user-images.githubusercontent.com/19517248/103180171-87c29200-4893-11eb-9f0c-823316fe554e.png)
After:
![image](https://user-images.githubusercontent.com/19517248/103180181-99a43500-4893-11eb-9bd6-1416f5cc434f.png)

mroavi committed:
Disable problematic query

mroavi commented:
This query is problematic:
```
((identifier) @type
  (match? @type "^[A-Z]"))
```
It captures variables that start with upper case and highlights them as types.
I disabled it but still need to find a replacement.

Before:
![image](https://user-images.githubusercontent.com/19517248/103179417-2e0a9980-488c-11eb-86cb-1a8b17ced7c0.png)
After:
![image](https://user-images.githubusercontent.com/19517248/103180331-26032780-4895-11eb-845b-8ff5cc2e2593.png)

mroavi committed:
Remove query that accidentaly got committed

mroavi commented:
@theHamsta mentioned that there are no "real priorities" between different queries. Can somebody explain to me how to get around this limitation? I have different queries that catch the same part of the code and I want to assign which one has more priority than the other.

theHamsta commented:
```
((identifier) @type
  (match? @type "^[A-Z]"))
```
What's the problem with that? We also use it in other languages. Isn't using first letter upper case also a bad thing in Julia? We have to follow style conventions when we're not having semantic analysis activated yet. You may add an exception for those one letter variables `A_foo`. Usually we put those naming convention queries on the top of files.

![image](https://user-images.githubusercontent.com/7189118/103181821-c06b6700-48a5-11eb-8327-8b6d60b5a914.png)

> I have different queries that catch the same part of the code and I want to assign which one has more priority than the other.

If it captures the same thing, the later one wins. If that doesn't work, you need to make one of the queries more precise but don't overdo it. After all it's only syntax highlighting based on the grammar without the additional precision that semantic analysis can give you

I actually meant that you create a PR at nvim-treesitter. But I can also merge your changes to my branch.

mroavi committed:
Add for_clause query

mroavi commented:
Before:
![image](https://user-images.githubusercontent.com/19517248/103182116-1db4e780-48a9-11eb-92dd-21fb6109df1c.png)
After:
![image](https://user-images.githubusercontent.com/19517248/103182124-31604e00-48a9-11eb-901b-fabbe8e8d225.png)

mroavi committed:
Re-enable "problematic" query, move it to the top

of the file. Add an exception for variables that start
with upper case followed by a dash.

mroavi committed:
Add query to capture parameterized_identifiers

mroavi commented:
Before:
![image](https://user-images.githubusercontent.com/19517248/103182504-0b887880-48ac-11eb-922b-ed6ca1efa398.png)
After:
![image](https://user-images.githubusercontent.com/19517248/103182511-193dfe00-48ac-11eb-973f-e11d98eb21c4.png)

theHamsta commented on queries/julia/highlights.scm:
```diff
@@ -1,4 +1,8 @@
+((identifier) @type
+  (match? @type "^[A-Z]"))
+((identifier) @variable
+  (match? @type "^[A-Z_]"))
```
This will not  work for most users since `@variable` will not have a highlight. I rather thought about  
```
((identifier) @type
  (match? @type "^[A-Z][a-Z0-9A-Z]"))
````
or 
```
((identifier) @type
  (match? @type "^[A-Z][^_]"))
```
Unfortunately, we don't have unicode regexes.

mroavi replied:
The first one gives an error. The second one works. I'll replace it with the second one.

mroavi committed:
Fix: error in predicate

mroavi committed:
Add comment explaning an exception

mroavi committed:
Replace @variable predicate

mroavi committed:
Add do_clause query

theHamsta commented:
Do you think this is ready for review for the main repo? We could defer the operators for a separated PR as they are very convoluted.

mroavi commented:
Yes. I think it is ready if we leave the operators for a separate PR.

clason commented:
@mroavi are you still around? would you be interested in helping with the parser itself?

mroavi commented:
Hi @clason. I'm afraid I don't have time at the moment. I'll reach out whenever I manage to free up some time.
